File: Swift/SymmetricTree.swift
//
//  LeetCode
//  Created by Abdelrahman Nasser
//

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func isSymmetric(_ root: TreeNode?) -> Bool {
          guard let root = root else {
            return true
        }
        
        return isMirror(root, root)
    }
    
    func isMirror(_ first: TreeNode?, _ second: TreeNode?) -> Bool {
         guard let firstVal = first?.val, let secondVal = second?.val else {
            return first == nil && second == nil
        }
        
        return (firstVal == secondVal) &&
        isMirror(first?.right, second?.left) &&
        isMirror(first?.left, second?.right)
    }
}

File: Java/SymmetricTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
  if (root==null ||root.left == null && root.right == null) return true;
        TreeNode left = root.left;
        TreeNode right = root.right;
        return isMirror(left, right);
    }
    public boolean isMirror(TreeNode l, TreeNode r) {
        if (l == null && r == null) return true;
        else if (l==null || r==null) return false;
        else if (l.val != r.val) return false;
        return isMirror(l.left, r.right) && isMirror(l.right, r.left);
    }
}